File: Raven.Client.Lightweight/Document/RemoteBulkInsertOperation.cs
using System.Diagnostics;
using System.Net;

using Raven.Abstractions.Exceptions;
using Raven.Abstractions.Json;
using Raven.Abstractions.Util;
using Raven.Client.Connection.Async;

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using Raven.Abstractions.Data;

using Raven.Abstractions.Connection;
using Raven.Abstractions.Extensions;
using Raven.Client.Changes;
using Raven.Client.Connection;
using Raven.Client.Connection.Implementation;
using Raven.Client.Util;
using Raven.Imports.Newtonsoft.Json;

using Raven.Imports.Newtonsoft.Json.Bson;
using Raven.Json.Linq;
using System.IO.Compression;
using System.Net.Http;
using Raven.Client.Extensions;
using System.Text;

namespace Raven.Client.Document
{

    public interface ILowLevelBulkInsertOperation : IDisposable
    {
        Guid OperationId { get; }

        bool IsAborted { get; }

        void Write(string id, RavenJObject metadata, RavenJObject data, int? dataSize = null);

        Task WriteAsync(string id, RavenJObject metadata, RavenJObject data, int? dataSize = null);

        Task WaitForLastTaskToFinish();

        Task<int> DisposeAsync();

        /// <summary>
        ///     Report on the progress of the operation
        /// </summary>
        event Action<string> Report;
        void Abort();
    }

    public class RemoteBulkInsertOperation : ILowLevelBulkInsertOperation, IObserver<BulkInsertChangeNotification>
    {
        private readonly BulkInsertOptions options;
        private readonly Task<int> previousTask;
        private CancellationTokenSource cancellationTokenSource;
        private readonly AsyncServerClient operationClient;
        private readonly MemoryStream bufferedStream = new MemoryStream();
        private readonly BlockingCollection<RavenJObject> queue;
        private static readonly RavenJObject AbortMarker = new RavenJObject();
        private static readonly RavenJObject SkipMarker = new RavenJObject();
        private HttpJsonRequest operationRequest;
        private readonly Task operationTask;
        private bool aborted;
        private bool waitedForPreviousTask;
        private readonly Stopwatch _timing = Stopwatch.StartNew();
        private const int BigDocumentSize = 64 * 1024;

        public RemoteBulkInsertOperation(BulkInsertOptions options, AsyncServerClient client, IDatabaseChanges changes,
            Task<int> previousTask = null, Guid? existingOperationId = null)
        {
            this.options = options;
            this.previousTask = previousTask;
            using (NoSynchronizationContext.Scope())
            {
                OperationId = existingOperationId.HasValue ? existingOperationId.Value : Guid.NewGuid();
                operationClient = client;
                queue = new BlockingCollection<RavenJObject>(Math.Max(128, (options.BatchSize * 3) / 2));

                operationTask = StartBulkInsertAsync(options);

#if !MONO
                SubscribeToBulkInsertNotifications(changes);
#endif
            }
        }

        public int Total { get; set; }
        public int localCount;
        public long size;

#if !MONO
        private void SubscribeToBulkInsertNotifications(IDatabaseChanges changes)
        {
            subscription = changes
                .ForBulkInsert(OperationId)
                .Subscribe(this);
        }
#endif

        private HttpResponseMessage response;

        private async Task StartBulkInsertAsync(BulkInsertOptions options)
        {
            using (ConnectionOptions.Expect100Continue(operationClient.Url))
            {
                var operationUrl = CreateOperationUrl(options);
                string token;

                try
                {
                    token = await GetToken().ConfigureAwait(false);
                }
                catch (Exception e)
                {
                    queue.CompleteAdding();
                    throw new InvalidOperationException("Could not get token for bulk insert", e);
                }

                try
                {
                    token = await ValidateThatWeCanUseAuthenticateTokens(token).ConfigureAwait(false);
                }
                catch (Exception e)
                {
                    queue.CompleteAdding();
                    throw new InvalidOperationException("Could not authenticate token for bulk insert, if you are using ravendb in IIS make sure you have Anonymous Authentication enabled in the IIS configuration", e);
                }

                using (operationRequest = CreateOperationRequest(operationUrl, token))
                {
                    var cancellationToken = CreateCancellationToken();
                    response = await operationRequest.ExecuteRawRequestAsync((stream, source) => Task.Factory.StartNew(() =>
                    {
                        try
                        {
                            WriteQueueToServer(stream, options, cancellationToken);
                            source.TrySetResult(null);
                        }
                        catch (Exception e)
                        {
                            //we get a cancellation only if we receive a notification of BulkInsertError
                            //in that case we need to get the real error from the server using response.AssertNotFailingResponse()
                            if (cancellationToken.IsCancellationRequested)
                                source.TrySetResult(null);
                            else
                                source.TrySetException(e);
                        }
                        finally
                        {
                            queue.CompleteAdding();
                        }
                    }, CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default)).ConfigureAwait(false);

                    await response.AssertNotFailingResponse().ConfigureAwait(false);

                    long operationId;

                    using (response)
                    using (var stream = await response.Content.ReadAsStreamAsync().ConfigureAwait(false))
                    using (var streamReader = new StreamReader(stream))
                    {
                        var result = RavenJObject.Load(new JsonTextReader(streamReader));
                        operationId = result.Value<long>("OperationId");
                    }

                    if (await IsOperationCompleted(operationId).ConfigureAwait(false))
                        responseOperationId = operationId;
                }
            }
        }

        private CancellationToken CreateCancellationToken()
        {
            cancellationTokenSource = new CancellationTokenSource();
            return cancellationTokenSource.Token;
        }

        private async Task<string> GetToken()
        {
            // this will force the HTTP layer to authenticate, meaning that our next request won't have to
            var jsonToken = await GetAuthToken().ConfigureAwait(false);

            return jsonToken.Value<string>("Token");
        }

        private async Task<RavenJToken> GetAuthToken()
        {
            using (var request = operationClient.CreateRequest("/singleAuthToken", HttpMethods.Get, disableRequestCompression: true))
            {
                return await request.ReadResponseJsonAsync().ConfigureAwait(false);
            }
        }

        private async Task<string> ValidateThatWeCanUseAuthenticateTokens(string token)
        {
            using (var request = operationClient.CreateRequest("/singleAuthToken", HttpMethods.Get, disableRequestCompression: true, disableAuthentication: true))
            {
                request.AddOperationHeader("Single-Use-Auth-Token", token);
                var result = await request.ReadResponseJsonAsync().ConfigureAwait(false);
                return result.Value<string>("Token");
            }
        }

        private HttpJsonRequest CreateOperationRequest(string operationUrl, string token)
        {
            // the request may take a long time to process, so we need to set a large timeout value
            var request = operationClient.CreateRequest(operationUrl, HttpMethods.Post, disableRequestCompression: true, disableAuthentication: true, timeout: TimeSpan.FromHours(6));
            request.AddOperationHeader("Single-Use-Auth-Token", token);

            return request;
        }

        private string CreateOperationUrl(BulkInsertOptions options)
        {
            var requestUrl = new StringBuilder();

            requestUrl.Append("/bulkInsert?");
            if (options.OverwriteExisting)
                requestUrl.Append("overwriteExisting=true");
            if (options.CheckReferencesInIndexes)
                requestUrl.Append("&checkReferencesInIndexes=true");
            if (options.SkipOverwriteIfUnchanged)
                requestUrl.Append("&skipOverwriteIfUnchanged=true");

            switch(options.Format)
            {
                case BulkInsertFormat.Bson: requestUrl.Append("&format=bson"); break;
                case BulkInsertFormat.Json: requestUrl.Append("&format=json"); break;
            }

            switch(options.Compression)
            {
                case BulkInsertCompression.None: requestUrl.Append("&compression=none"); break;
                case BulkInsertCompression.GZip: requestUrl.Append("&compression=gzip"); break;
        }

            requestUrl.Append("&operationId=" + OperationId);

            return requestUrl.ToString();
        }

        private void WriteQueueToServer(Stream stream, BulkInsertOptions options, CancellationToken cancellationToken)
        {
            while (true)
            {
                cancellationToken.ThrowIfCancellationRequested();

                var batch = new List<RavenJObject>();
                RavenJObject document;
                var sp = Stopwatch.StartNew();
                while (queue.TryTake(out document, millisecondsTimeout: 200))
                {
                    cancellationToken.ThrowIfCancellationRequested();

                    if (document == null) // marker
                    {
                        FlushBatch(stream, batch);
                        return;
                    }
                    if (ReferenceEquals(SkipMarker, document)) // ignore this, just filling the queue
                    {
                        continue;
                    }
                    if (ReferenceEquals(AbortMarker, document)) // abort immediately
                    {
                        return;
                    }

                    batch.Add(document);

                    if (batch.Count >= options.BatchSize)
                        break;

                    // force flush if we didn't do that in 10 seconds in a row
                    if (sp.ElapsedMilliseconds > 10*1000)
                        break;
                }

                FlushBatch(stream, batch);
            }
        }

        public event Action<string> Report;

        public Guid OperationId { get; private set; }

        public virtual void Write(string id, RavenJObject metadata, RavenJObject data, int? dataSize = null)
        {
            if (id == null) throw new ArgumentNullException("id");
            if (metadata == null) throw new ArgumentNullException("metadata");
            if (data == null) throw new ArgumentNullException("data");
            if (aborted) throw new InvalidOperationException("Operation has been aborted");


            metadata["@id"] = id;
            data[Constants.Metadata] = metadata;

            for (int i = 0; i < 2; i++)
            {
                if (operationTask.IsCanceled || operationTask.IsFaulted)
                    operationTask.Wait(); //error early if we have any error

                try
                {
                    if (queue.TryAdd(data, options.WriteTimeoutMilliseconds / 2))
                    {
                        if (dataSize != null && dataSize >= BigDocumentSize)
                        {
                            //essentially for a BatchSize == 1024 and stream of 1MB documents - the actual batch size will be 128
                            // --> BatchSize = 1024 / (dataSize = 1024/BigDocumentSize = 250) * 2 == 128
                            for (int skipDocIndex = 0; skipDocIndex < (dataSize / BigDocumentSize) * 2; skipDocIndex++)
                            {
                                if (queue.TryAdd(SkipMarker) == false) //if queue is full just stop adding dummy docs
                                    break;
                            }
                        }
                        return;
                    }
                }
                catch (InvalidOperationException e)
                {
                    //this means that the queue is marked as complete,
                    //probably, because there was an error

                    break;
                }
            }

            if (operationTask.IsCanceled || operationTask.IsFaulted)
                operationTask.Wait(); //error early if we have any error

            throw new TimeoutException("Could not flush in the specified timeout, server probably not responding or responding too slowly.\r\nAre you writing very big documents?");
        }

        public virtual async Task WriteAsync(string id, RavenJObject metadata, RavenJObject data, int? dataSize = null)
        {
            if (id == null) throw new ArgumentNullException("id");
            if (metadata == null) throw new ArgumentNullException("metadata");
            if (data == null) throw new ArgumentNullException("data");
            if (aborted) throw new InvalidOperationException("Operation has been aborted");


            metadata["@id"] = id;
            data[Constants.Metadata] = metadata;

            for (int i = 0; i < 2; i++)
            {
                if (operationTask.IsCanceled || operationTask.IsFaulted)
                    await operationTask.ConfigureAwait(false); //error early if we have any error

                try
                {
                    if (queue.TryAdd(data, options.WriteTimeoutMilliseconds / 2))
                    {
                        if (dataSize != null && dataSize >= BigDocumentSize)
                        {
                            //essentially for a BatchSize == 1024 and stream of 1MB documents - the actual batch size will be 128
                            // --> BatchSize = 1024 / (dataSize = 1024/BigDocumentSize = 250) * 2 == 128
                            for (int skipDocIndex = 0; skipDocIndex < (dataSize / BigDocumentSize) * 2; skipDocIndex++)
                            {
                                if (queue.TryAdd(SkipMarker) == false) //if queue is full just stop adding dummy docs
                                    break;
                            }
                        }
                        return;
                    }
                }
                catch (InvalidOperationException e)
                {
                    //this means that the queue is marked as complete,
                    //probably, because there was an error

                    break;
                }
            }

            if (operationTask.IsCanceled || operationTask.IsFaulted)
                await operationTask.ConfigureAwait(false); //error early if we have any error

            throw new TimeoutException("Could not flush in the specified timeout, server probably not responding or responding too slowly.\r\nAre you writing very big documents?");
        }

        public async Task WaitForLastTaskToFinish()
        {
            throw new NotImplementedException("WaitForLastTaskToFinish not implemented for RemoteBulkInsertOperation, see ChunkedRemoteBulkInsertOperation");
        }

        private async Task<bool> IsOperationCompleted(long operationId)
        {
            ErrorResponseException errorResponse;

            try
            {
                var status = await GetOperationStatus(operationId).ConfigureAwait(false);

                if (status == null)
                    return true;

                if (status.Value<bool>("Completed"))
                    return true;

                return false;
            }
            catch (ErrorResponseException e)
            {
                if (e.StatusCode != HttpStatusCode.Conflict)
                    throw;

                errorResponse = e;
            }

            using(var stringReader = new StringReader(errorResponse.ResponseString))
            using (var ravenJsonTextReader = new RavenJsonTextReader(stringReader))
            {
                var conflictsDocument = RavenJObject.Load(ravenJsonTextReader);

                throw new ConcurrencyException(conflictsDocument.Value<string>("Error"));
            }
        }

        private Task<RavenJToken> GetOperationStatus(long operationId)
        {
            return operationClient.GetOperationStatusAsync(operationId);
        }

        private volatile bool disposed;
        private IDisposable subscription;

        private long responseOperationId;

        public async Task<int> DisposeAsync()
        {
            if (disposed)
                return -1;
            disposed = true;

            try
            {
                // adding the "finished" marker to avoid an infinite loop
                queue.Add(null);
            }
            catch (InvalidOperationException e)
            {
                // this means that the queue is marked as complete,
                // probably, there was an error and we'll get the reason for it
                // when running "operationTask.AssertNotFailed()" later on
            }

            if (subscription != null)
            {
                subscription.Dispose();
            }

            // The first await call in this method MUST call ConfigureAwait(false) in order to avoid DEADLOCK when this code is called by synchronize code, like Dispose().
            try
            {
                await operationTask.ConfigureAwait(false);
#pragma warning disable 4014
                operationTask.AssertNotFailed();
#pragma warning restore 4014

                if (previousTask == null)
                    ReportInternal("Finished writing all results to server");

                while (true)
                {
                    if (await IsOperationCompleted(responseOperationId).ConfigureAwait(false))
                        break;

                    await Task.Delay(100).ConfigureAwait(false);
                }
                if (previousTask == null)
                {
                    ReportInternal("Done writing to server");
                }
                else
                {
                    ReportInternal("Wrote {0:#,#} [{3:#,#;;0} kb] (total {2:#,#;;0}) documents to server gzipped to {1:#,#;;0} kb in {4:#,#.#;;0} sec.",
                       localCount,
                       bufferedStream.Position / 1024d,
                       Total,
                       size / 1024d,
                       _timing.Elapsed.TotalSeconds);
                }
            }
            catch (Exception e)
            {
                ReportInternal("Failed to write all results to a server, probably something happened to the server. Exception : {0}", e);
                if (e.Message.Contains("Raven.Abstractions.Exceptions.ConcurrencyException"))
                    throw new ConcurrencyException("ConcurrencyException while writing bulk insert items in the server. Did you run bulk insert operation with OverwriteExisting == false?. Exception returned from server: " + e.Message, e);

                if (e.Message.Contains("Raven.Abstractions.Exceptions.OperationVetoedException"))
                    throw new OperationVetoedException(e.Message, e);
                throw;
            }
            return Total;
        }

        public void Dispose()
        {
            if (disposed)
                return;

            using (NoSynchronizationContext.Scope())
            {
                var disposeAsync = DisposeAsync().ConfigureAwait(false);
                disposeAsync.GetAwaiter().GetResult();
            }
        }

        private void FlushBatch(Stream requestStream, ICollection<RavenJObject> localBatch)
        {
            if (localBatch.Count == 0)
                return;
            if (aborted) throw new InvalidOperationException("Operation was timed out or has been aborted");

            if (previousTask != null && waitedForPreviousTask == false)
            {
                Total += previousTask.Result;
                waitedForPreviousTask = true;
            }

            var sp = Stopwatch.StartNew();

            bufferedStream.SetLength(0);
            long bytesWrittenToServer = WriteToBuffer(options, bufferedStream, localBatch);

            var requestBinaryWriter = new BinaryWriter(requestStream);
            requestBinaryWriter.Write((int)bufferedStream.Position);
            bufferedStream.WriteTo(requestStream);
            requestStream.Flush();

            Total += localBatch.Count;
            localCount += localBatch.Count;
            size += bytesWrittenToServer;

            if (previousTask == null)
            {
                ReportInternal("Wrote {0:#,#} [{3:#,#;;0} kb] (total {2:#,#;;0}) documents to server gzipped to {1:#,#;;0} kb in {4:#,#.#;;0} sec.",
                    localBatch.Count,
                    bufferedStream.Position / 1024d,
                    Total,
                    bytesWrittenToServer / 1024d,
                    sp.Elapsed.TotalSeconds);
            }
        }

        private static long WriteToBuffer(BulkInsertOptions options, Stream stream, ICollection<RavenJObject> batch)
        {
            switch (options.Compression)
            {
                case BulkInsertCompression.GZip:
                {
                    using (var gzip = new GZipStream(stream, CompressionMode.Compress, leaveOpen: true))
                    {
                        return WriteBatchToBuffer(options, gzip, batch);
                    }
                }
                case BulkInsertCompression.None:
                {
                    return WriteBatchToBuffer(options, stream, batch);
                }
                default: throw new NotSupportedException(string.Format("The compression algorithm '{0}' is not supported", options.Compression.ToString()));
            }
        }

        private static long WriteBatchToBuffer(BulkInsertOptions options, Stream stream, ICollection<RavenJObject> batch)
        {
            using (var countingStream = new CountingStream(stream))
            {
                switch (options.Format)
                {
                    case BulkInsertFormat.Bson:
                    {
                        WriteBsonBatchToBuffer(options, countingStream, batch);
                        break;
                    }
                    case BulkInsertFormat.Json:
                    {
                        WriteJsonBatchToBuffer(options, countingStream, batch);
                        break;
                    }
                    default: throw new NotSupportedException(string.Format("The format '{0}' is not supported", options.Format.ToString()));
                }

                countingStream.Flush();
                return countingStream.NumberOfWrittenBytes;
            }
        }

        private static void WriteBsonBatchToBuffer(BulkInsertOptions options, CountingStream stream, ICollection<RavenJObject> batch)
        {
            var binaryWriter = new BinaryWriter(stream);
            binaryWriter.Write(batch.Count);
            binaryWriter.Flush();

            var bsonWriter = new BsonWriter(binaryWriter)
            {
                DateTimeKindHandling = DateTimeKind.Unspecified
            };

            foreach (var doc in batch)
            {
                doc.WriteTo(bsonWriter);
            }

            bsonWriter.Flush();
        }

        private static void WriteJsonBatchToBuffer(BulkInsertOptions options, CountingStream stream, ICollection<RavenJObject> batch)
        {
            var binaryWriter = new BinaryWriter(stream);
            binaryWriter.Write(batch.Count);
            binaryWriter.Flush();

            var jsonWriter = new JsonTextWriter(new StreamWriter(stream))
            {
                DateFormatHandling = DateFormatHandling.IsoDateFormat
            };

            foreach (var doc in batch)
            {
                doc.WriteTo(jsonWriter);
            }

            jsonWriter.Flush();
        }

        private void ReportInternal(string format, params object[] args)
        {
            var onReport = Report;
            if (onReport != null)
                onReport(string.Format(format, args));
        }

        public void OnNext(BulkInsertChangeNotification value)
        {
            if (value.Type == DocumentChangeTypes.BulkInsertError)
            {
                cancellationTokenSource.Cancel();
            }
        }

        public void OnError(Exception error)
        {
        }

        public void OnCompleted()
        {
        }

        public void Abort()
        {
            aborted = true;
            queue.Add(AbortMarker);
        }


        public bool IsAborted
        {
            get { return aborted; }
        }
    }
}

File: Raven.Client.Java/src/main/java/net/ravendb/client/document/RemoteBulkInsertOperation.java
package net.ravendb.client.document;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.lang.Thread.UncaughtExceptionHandler;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.TimeUnit;
import java.util.zip.GZIPOutputStream;

import net.ravendb.abstractions.closure.Action1;
import net.ravendb.abstractions.data.BulkInsertChangeNotification;
import net.ravendb.abstractions.data.BulkInsertOptions;
import net.ravendb.abstractions.data.Constants;
import net.ravendb.abstractions.data.DocumentChangeTypes;
import net.ravendb.abstractions.data.HttpMethods;
import net.ravendb.abstractions.json.linq.RavenJObject;
import net.ravendb.abstractions.json.linq.RavenJToken;
import net.ravendb.client.changes.IDatabaseChanges;
import net.ravendb.client.changes.IObserver;
import net.ravendb.client.connection.ServerClient;
import net.ravendb.client.connection.implementation.HttpJsonRequest;
import net.ravendb.client.utils.CancellationTokenSource;
import net.ravendb.client.utils.CancellationTokenSource.CancellationToken;

import org.apache.commons.lang.ArrayUtils;
import org.apache.http.Header;
import org.apache.http.HttpEntity;
import org.apache.http.client.methods.HttpPost;

import de.undercouch.bson4jackson.BsonFactory;
import de.undercouch.bson4jackson.BsonGenerator;


public class RemoteBulkInsertOperation implements ILowLevelBulkInsertOperation, IObserver<BulkInsertChangeNotification> {

  private final BsonFactory bsonFactory = new BsonFactory();

  private final static RavenJObject END_OF_QUEUE_OBJECT = RavenJObject.parse("{ \"QueueFinished\" : true }");

  private CancellationTokenSource cancellationTokenSource;
  private final ServerClient operationClient;

  private final ByteArrayOutputStream bufferedStream = new ByteArrayOutputStream();
  private final BlockingQueue<RavenJObject> queue;

  private HttpJsonRequest operationRequest;
  private byte[] responseBytes;
  private final Thread operationTask;
  private Exception operationTaskException;
  private int total;

  private Action1<String> report;
  private UUID operationId;
  private transient boolean disposed;

  @Override
  public UUID getOperationId() {
    return operationId;
  }

  @Override
  public Action1<String> getReport() {
    return report;
  }

  @Override
  public void setReport(Action1<String> report) {
    this.report = report;
  }



  public RemoteBulkInsertOperation(BulkInsertOptions options, ServerClient client, IDatabaseChanges changes) {
    operationId = UUID.randomUUID();
    operationClient = client;
    queue = new ArrayBlockingQueue<>(Math.max(128, (options.getBatchSize() * 3) / 2));

    operationTask = startBulkInsertAsync(options);
    subscribeToBulkInsertNotifications(changes);
  }

  private void subscribeToBulkInsertNotifications(IDatabaseChanges changes) {
    changes.forBulkInsert(operationId).subscribe(this);
  }

  private class BulkInsertEntity implements HttpEntity {

    private BulkInsertOptions options;
    private CancellationToken cancellationToken;

    public BulkInsertEntity(BulkInsertOptions options, CancellationToken cancellationToken) {
      this.options = options;
      this.cancellationToken = cancellationToken;
    }

    @Override
    public boolean isRepeatable() {
      return false;
    }

    @Override
    public boolean isChunked() {
      return true;
    }

    @Override
    public long getContentLength() {
      return 0;
    }

    @Override
    public Header getContentType() {
      return null;
    }

    @Override
    public Header getContentEncoding() {
      return null;
    }

    @Override
    public InputStream getContent() throws IOException, IllegalStateException {
      throw new IllegalStateException("Not supported!");
    }

    @Override
    public void writeTo(OutputStream outstream) throws IOException {
      writeQueueToServer(outstream, options, cancellationToken);
    }

    @Override
    public boolean isStreaming() {
      return true;
    }

    @Deprecated
    @Override
    public void consumeContent() throws IOException {
      //empty
    }

  }

  private CancellationToken createCancellationToken() {
    cancellationTokenSource = new CancellationTokenSource();
    return cancellationTokenSource.getToken();
  }

  private Thread startBulkInsertAsync(BulkInsertOptions options) {
    operationClient.setExpect100Continue(true);

    String operationUrl = createOperationUrl(options);
    String token = getToken();
    try {
      token = validateThatWeCanUseAuthenticateTokens(token);
    } catch (Exception e) {
      throw new IllegalStateException("Could not authenticate token for bulk insert, if you are using ravendb in IIS make sure you have Anonymous Authentication enabled in the IIS configuration", e);
    }

    operationRequest = createOperationRequest(operationUrl, token);
    HttpPost webRequest = (HttpPost) operationRequest.getWebRequest();
    webRequest.setEntity(new BulkInsertEntity(options, createCancellationToken()));


    Thread thread = new Thread(new Runnable() {

      @Override
      public void run() {
        try {
          responseBytes = operationRequest.readResponseBytes();
        } catch (IOException e) {
          throw new RuntimeException(e);
        }
      }
    });

    operationClient.setExpect100Continue(false);
    thread.setUncaughtExceptionHandler(new UncaughtExceptionHandler() {
      @Override
      public void uncaughtException(Thread t, Throwable e) {
        operationTaskException = (Exception) e;
      }
    });

    thread.start();
    return thread;

  }

  private String getToken() {
    RavenJToken jsonToken = getAuthToken();
    return jsonToken.value(String.class, "Token");
  }

  private RavenJToken getAuthToken() {
    HttpJsonRequest request = operationClient.createRequest(HttpMethods.GET, "/singleAuthToken", true);
    return request.readResponseJson();
  }

  private String validateThatWeCanUseAuthenticateTokens(String token) {
    HttpJsonRequest request = operationClient.createRequest(HttpMethods.GET, "/singleAuthToken", true);
    request.removeAuthorizationHeader();
    request.addOperationHeader("Single-Use-Auth-Token", token);
    RavenJToken result = request.readResponseJson();
    return result.value(String.class, "Token");
  }

  private HttpJsonRequest createOperationRequest(String operationUrl, String token) {
    HttpJsonRequest request = operationClient.createRequest(HttpMethods.POST, operationUrl, true);

    request.prepareForLongRequest();
    request.addOperationHeader("Single-Use-Auth-Token", token);

    return request;
  }

  private String createOperationUrl(BulkInsertOptions options) {
    String requestUrl = "/bulkInsert?";
    if (options.isCheckForUpdates()) {
      requestUrl += "checkForUpdates=true";
    }
    if (options.isCheckReferencesInIndexes()) {
      requestUrl += "&checkReferencesInIndexes=true";
    }

    requestUrl += "&operationId=" + operationId;

    return requestUrl;
  }
  private void writeQueueToServer(OutputStream stream, BulkInsertOptions options, CancellationToken cancellationToken) throws IOException {

    while (true) {
      cancellationToken.throwIfCancellationRequested();
      List<RavenJObject> batch = new ArrayList<>();
      try {
        RavenJObject document;
        while ((document = queue.poll(200, TimeUnit.MICROSECONDS)) != null) {
          cancellationToken.throwIfCancellationRequested();

          if (document == END_OF_QUEUE_OBJECT) { //marker
            flushBatch(stream, batch);
            return;
          }
          batch.add(document);

          if (batch.size() >= options.getBatchSize()) {
            break;
          }
        }
      } catch (InterruptedException e ){
        //ignore
      }
      flushBatch(stream, batch);
    }
  }

  @Override
  public void write(String id, RavenJObject metadata, RavenJObject data) {
    if (id == null) {
      throw new IllegalArgumentException("id");
    }
    if (metadata == null) {
      throw new IllegalArgumentException("metadata");
    }
    if (data == null) {
      throw new IllegalArgumentException("data");
    }

    if (operationTaskException != null) {
      throw new RuntimeException(operationTaskException); // error early if we have  any error
    }

    metadata.add("@id", id);
    data.add(Constants.METADATA, metadata);
    try {
      while (!queue.offer(data)) {
        Thread.sleep(250);
      }
    } catch (InterruptedException e) {
      throw new RuntimeException(e);
    }
  }


  private boolean isOperationCompleted(long operationId) {
    RavenJToken status = getOperationStatus(operationId);
    if (status == null) {
      return true;
    }
    if (status.value(Boolean.class, "Completed")) {
      return true;
    }
    return false;
  }

  private RavenJToken getOperationStatus(long operationId) {
    return operationClient.getOperationStatus(operationId);
  }


  @Override
  public void close() throws InterruptedException {
    if (disposed) {
      return ;
    }
    queue.add(END_OF_QUEUE_OBJECT);
    operationTask.join();

    if (operationTaskException != null) {
      throw new RuntimeException(operationTaskException);
    }

    reportInternal("Finished writing all results to server");

    long operationId = 0;

    RavenJObject result = RavenJObject.parse(new String(responseBytes));
    operationId = result.value(Long.class, "OperationId");

    while (true) {
      if (isOperationCompleted(operationId)) {
        break;
      }
      Thread.sleep(500);
    }
    reportInternal("Done writing to server");
  }

  private  void flushBatch(OutputStream requestStream, Collection<RavenJObject> localBatch) throws IOException {
    if (localBatch.isEmpty()) {
      return ;
    }
    bufferedStream.reset();
    writeToBuffer(localBatch);

    byte[] bytes = ByteBuffer.allocate(4).putInt(bufferedStream.size()).array();
    ArrayUtils.reverse(bytes);
    requestStream.write(bytes);
    bufferedStream.writeTo(requestStream);
    requestStream.flush();

    total += localBatch.size();

    Action1<String> report = getReport();
    if (report != null) {
      report.apply(String.format("Wrote %d (total %d) documents to server gzipped to %d kb", localBatch.size(), total, bufferedStream.size() / 1024));
    }

  }

  private void writeToBuffer(Collection<RavenJObject> localBatch) throws IOException {
    GZIPOutputStream gzipOutputStream = new GZIPOutputStream(bufferedStream);

    BsonGenerator bsonWriter = bsonFactory.createJsonGenerator(gzipOutputStream);
    bsonWriter.disable(org.codehaus.jackson.JsonGenerator.Feature.AUTO_CLOSE_TARGET);

    byte[] bytes = ByteBuffer.allocate(4).putInt(localBatch.size()).array();
    ArrayUtils.reverse(bytes);
    gzipOutputStream.write(bytes);
    for (RavenJObject doc : localBatch) {
      doc.writeTo(bsonWriter);
    }
    bsonWriter.close();
    gzipOutputStream.finish();
    bufferedStream.flush();
  }

  private void reportInternal(String format, Object... args) {
    Action1<String> onReport = report;
    if (onReport != null) {
      onReport.apply(String.format(format, args));
    }
  }

  @Override
  public void onNext(BulkInsertChangeNotification value) {
    if (value.getType().equals(DocumentChangeTypes.BULK_INSERT_ERROR)) {
      cancellationTokenSource.cancel();
    }
  }

  @Override
  public void onError(Exception error) {
    //empty by design
  }

  @Override
  public void onCompleted() {
    //empty by design
  }
}
